File: cc-native-compiler-osaka-202605/langs/typescript/src/typecheck.ts
import type {
  Expr, FunDef, Program, Stmt, Type,
} from "./ast.ts";
import { typeToString } from "./ast.ts";

export class TypeError extends Error {
  constructor(msg: string) {
    super(msg);
    this.name = "TypeError";
  }
}

type VarInfo = { ty: Type; isMutable: boolean };
type FunSig  = { params: Type[]; retType: Type };

export function check(program: Program): void {
  const funcs = new Map<string, FunSig>();
  for (const f of program.funcs) {
    funcs.set(f.name, { params: f.params.map((p) => p.ty), retType: f.retType });
  }

  for (const f of program.funcs) {
    const env = new Map<string, VarInfo>();
    for (const p of f.params) {
      env.set(p.name, { ty: p.ty, isMutable: false });
    }
    checkStmts(f.body, env, funcs, f.retType);
  }

  checkStmts(program.topStmts, new Map(), funcs, null);
}

function checkStmts(
  stmts: Stmt[],
  envIn: Map<string, VarInfo>,
  funcs: Map<string, FunSig>,
  retTy: Type | null,
): Map<string, VarInfo> {
  let env = new Map(envIn);
  for (const s of stmts) {
    env = checkStmt(s, env, funcs, retTy);
  }
  return env;
}

function checkStmt(
  stmt: Stmt,
  env: Map<string, VarInfo>,
  funcs: Map<string, FunSig>,
  retTy: Type | null,
): Map<string, VarInfo> {
  switch (stmt.kind) {
    case "ValDef": {
      const initTy = checkExpr(stmt.init, env, funcs);
      let finalTy = stmt.declared;
      if (finalTy === null) {
        finalTy = initTy;
      } else if (!sameType(finalTy, initTy)) {
        throw new TypeError(
          `val ${stmt.name}: declared ${typeToString(finalTy)} but init is ${typeToString(initTy)}`,
        );
      }
      const next = new Map(env);
      next.set(stmt.name, { ty: finalTy, isMutable: false });
      return next;
    }
    case "VarDef": {
      const initTy = checkExpr(stmt.init, env, funcs);
      let finalTy = stmt.declared;
      if (finalTy === null) {
        finalTy = initTy;
      } else if (!sameType(finalTy, initTy)) {
        throw new TypeError(
          `var ${stmt.name}: declared ${typeToString(finalTy)} but init is ${typeToString(initTy)}`,
        );
      }
      const next = new Map(env);
      next.set(stmt.name, { ty: finalTy, isMutable: true });
      return next;
    }
    case "Reassign": {
      const info = env.get(stmt.name);
      if (!info) throw new TypeError(`undefined variable: ${stmt.name}`);
      if (!info.isMutable) throw new TypeError(`cannot reassign val: ${stmt.name}`);
      const vt = checkExpr(stmt.value, env, funcs);
      if (!sameType(info.ty, vt)) {
        throw new TypeError(
          `${stmt.name}: type ${typeToString(info.ty)} but got ${typeToString(vt)}`,
        );
      }
      return env;
    }
    case "If": {
      const ct = checkExpr(stmt.cond, env, funcs);
      if (ct.kind !== "TInt") {
        throw new TypeError(`if condition must be int (bool), got ${typeToString(ct)}`);
      }
      checkStmts(stmt.thenBlk, env, funcs, retTy);
      checkStmts(stmt.elseBlk, env, funcs, retTy);
      return env;
    }
    case "While": {
      const ct = checkExpr(stmt.cond, env, funcs);
      if (ct.kind !== "TInt") {
        throw new TypeError(`while condition must be int (bool), got ${typeToString(ct)}`);
      }
      checkStmts(stmt.body, env, funcs, retTy);
      return env;
    }
    case "Print":
      checkExpr(stmt.value, env, funcs);
      return env;
    case "Return": {
      const vt = checkExpr(stmt.value, env, funcs);
      if (retTy === null) throw new TypeError("return outside function");
      if (!sameType(retTy, vt)) {
        throw new TypeError(
          `return type mismatch: expected ${typeToString(retTy)}, got ${typeToString(vt)}`,
        );
      }
      return env;
    }
    case "ExprStmt":
      checkExpr(stmt.value, env, funcs);
      return env;
  }
}

function checkExpr(
  e: Expr,
  env: Map<string, VarInfo>,
  funcs: Map<string, FunSig>,
): Type {
  switch (e.kind) {
    case "IntLit": return { kind: "TInt" };
    case "StrLit": return { kind: "TString" };
    case "Var": {
      const info = env.get(e.name);
      if (!info) throw new TypeError(`undefined variable: ${e.name}`);
      return info.ty;
    }
    case "Neg": {
      const t = checkExpr(e.inner, env, funcs);
      if (t.kind !== "TInt") throw new TypeError(`negation requires int, got ${typeToString(t)}`);
      return { kind: "TInt" };
    }
    case "BinOp": {
      const lt = checkExpr(e.left, env, funcs);
      const rt = checkExpr(e.right, env, funcs);
      if (!sameType(lt, rt)) {
        throw new TypeError(`binop ${e.op}: type mismatch ${typeToString(lt)} vs ${typeToString(rt)}`);
      }
      switch (e.op) {
        case "+": case "-": case "*": case "/": case "%":
          if (lt.kind !== "TInt") {
            throw new TypeError(`arithmetic ${e.op} requires int, got ${typeToString(lt)}`);
          }
          return { kind: "TInt" };
        case "==": case "!=": case "<": case "<=": case ">": case ">=":
          if (lt.kind !== "TInt") {
            throw new TypeError(`comparison ${e.op} requires int, got ${typeToString(lt)}`);
          }
          return { kind: "TInt" };
        default:
          throw new TypeError(`unknown op: ${e.op}`);
      }
    }
    case "Call": {
      const sig = funcs.get(e.name);
      if (!sig) throw new TypeError(`undefined function: ${e.name}`);
      if (e.args.length !== sig.params.length) {
        throw new TypeError(
          `${e.name}: expected ${sig.params.length} args, got ${e.args.length}`,
        );
      }
      for (let i = 0; i < e.args.length; i++) {
        const at = checkExpr(e.args[i]!, env, funcs);
        const pt = sig.params[i]!;
        if (!sameType(at, pt)) {
          throw new TypeError(
            `${e.name}: arg type ${typeToString(at)}, expected ${typeToString(pt)}`,
          );
        }
      }
      return sig.retType;
    }
    case "Index": {
      const at = checkExpr(e.arr, env, funcs);
      const it = checkExpr(e.idx, env, funcs);
      if (it.kind !== "TInt") throw new TypeError(`index must be int, got ${typeToString(it)}`);
      if (at.kind === "TList") return at.elem;
      throw new TypeError(`cannot index non-list type: ${typeToString(at)}`);
    }
    case "ListLit": {
      if (e.elems.length === 0) throw new TypeError("empty list literal is not supported");
      const first = checkExpr(e.elems[0]!, env, funcs);
      for (let i = 1; i < e.elems.length; i++) {
        const t = checkExpr(e.elems[i]!, env, funcs);
        if (!sameType(t, first)) {
          throw new TypeError(
            `list literal: mixed types ${typeToString(first)} and ${typeToString(t)}`,
          );
        }
      }
      return { kind: "TList", elem: first };
    }
  }
}

function sameType(a: Type, b: Type): boolean {
  if (a.kind === "TInt"    && b.kind === "TInt")    return true;
  if (a.kind === "TString" && b.kind === "TString") return true;
  if (a.kind === "TList"   && b.kind === "TList")   return sameType(a.elem, b.elem);
  return false;
}

File: cc-native-compiler-osaka-202605/langs/java/src/main/java/nblang/TypeCheck.java
package nblang;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import nblang.Ast.Expr;
import nblang.Ast.FunDef;
import nblang.Ast.Param;
import nblang.Ast.Program;
import nblang.Ast.Stmt;
import nblang.Ast.Type;

public final class TypeCheck {
    public static final class TypeError extends RuntimeException {
        public TypeError(String msg) { super(msg); }
    }

    private record VarInfo(Type ty, boolean isMutable) {}
    private record FunSig(List<Type> params, Type retType) {}

    private TypeCheck() {}

    public static void check(Program program) {
        Map<String, FunSig> funcs = new HashMap<>();
        for (FunDef f : program.funcs()) {
            List<Type> paramTypes = f.params().stream().map(Param::ty).toList();
            funcs.put(f.name(), new FunSig(paramTypes, f.retType()));
        }

        for (FunDef f : program.funcs()) {
            Map<String, VarInfo> env = new HashMap<>();
            for (Param p : f.params()) {
                env.put(p.name(), new VarInfo(p.ty(), false));
            }
            checkStmts(f.body(), env, funcs, f.retType());
        }

        checkStmts(program.topStmts(), new HashMap<>(), funcs, null);
    }

    private static Map<String, VarInfo> checkStmts(
            List<Stmt> stmts,
            Map<String, VarInfo> envIn,
            Map<String, FunSig> funcs,
            Type retTy) {
        Map<String, VarInfo> env = new HashMap<>(envIn);
        for (Stmt s : stmts) {
            env = checkStmt(s, env, funcs, retTy);
        }
        return env;
    }

    private static Map<String, VarInfo> checkStmt(
            Stmt stmt,
            Map<String, VarInfo> env,
            Map<String, FunSig> funcs,
            Type retTy) {
        return switch (stmt) {
            case Stmt.ValDef v -> {
                Type initTy = checkExpr(v.init(), env, funcs);
                Type finalTy = v.declared();
                if (finalTy == null) {
                    finalTy = initTy;
                } else if (!sameType(finalTy, initTy)) {
                    throw new TypeError("val " + v.name() + ": declared " + finalTy + " but init is " + initTy);
                }
                Map<String, VarInfo> next = new HashMap<>(env);
                next.put(v.name(), new VarInfo(finalTy, false));
                yield next;
            }
            case Stmt.VarDef v -> {
                Type initTy = checkExpr(v.init(), env, funcs);
                Type finalTy = v.declared();
                if (finalTy == null) {
                    finalTy = initTy;
                } else if (!sameType(finalTy, initTy)) {
                    throw new TypeError("var " + v.name() + ": declared " + finalTy + " but init is " + initTy);
                }
                Map<String, VarInfo> next = new HashMap<>(env);
                next.put(v.name(), new VarInfo(finalTy, true));
                yield next;
            }
            case Stmt.Reassign r -> {
                VarInfo info = env.get(r.name());
                if (info == null) throw new TypeError("undefined variable: " + r.name());
                if (!info.isMutable()) throw new TypeError("cannot reassign val: " + r.name());
                Type vt = checkExpr(r.value(), env, funcs);
                if (!sameType(info.ty(), vt)) {
                    throw new TypeError(r.name() + ": type " + info.ty() + " but got " + vt);
                }
                yield env;
            }
            case Stmt.IfS i -> {
                Type ct = checkExpr(i.cond(), env, funcs);
                if (!(ct instanceof Type.TInt)) {
                    throw new TypeError("if condition must be int (bool), got " + ct);
                }
                checkStmts(i.thenBlk(), env, funcs, retTy);
                checkStmts(i.elseBlk(), env, funcs, retTy);
                yield env;
            }
            case Stmt.WhileS w -> {
                Type ct = checkExpr(w.cond(), env, funcs);
                if (!(ct instanceof Type.TInt)) {
                    throw new TypeError("while condition must be int (bool), got " + ct);
                }
                checkStmts(w.body(), env, funcs, retTy);
                yield env;
            }
            case Stmt.Print p -> {
                checkExpr(p.value(), env, funcs);
                yield env;
            }
            case Stmt.Return r -> {
                Type vt = checkExpr(r.value(), env, funcs);
                if (retTy == null) throw new TypeError("return outside function");
                if (!sameType(retTy, vt)) {
                    throw new TypeError("return type mismatch: expected " + retTy + ", got " + vt);
                }
                yield env;
            }
            case Stmt.ExprStmt es -> {
                checkExpr(es.value(), env, funcs);
                yield env;
            }
        };
    }

    private static Type checkExpr(Expr e, Map<String, VarInfo> env, Map<String, FunSig> funcs) {
        return switch (e) {
            case Expr.IntLit ignored -> new Type.TInt();
            case Expr.StrLit ignored -> new Type.TString();
            case Expr.Var v -> {
                VarInfo info = env.get(v.name());
                if (info == null) throw new TypeError("undefined variable: " + v.name());
                yield info.ty();
            }
            case Expr.Neg neg -> {
                Type t = checkExpr(neg.inner(), env, funcs);
                if (!(t instanceof Type.TInt)) throw new TypeError("negation requires int, got " + t);
                yield new Type.TInt();
            }
            case Expr.BinOp b -> {
                Type lt = checkExpr(b.left(), env, funcs);
                Type rt = checkExpr(b.right(), env, funcs);
                if (!sameType(lt, rt)) {
                    throw new TypeError("binop " + b.op() + ": type mismatch " + lt + " vs " + rt);
                }
                yield switch (b.op()) {
                    case "+", "-", "*", "/", "%" -> {
                        if (!(lt instanceof Type.TInt)) {
                            throw new TypeError("arithmetic " + b.op() + " requires int, got " + lt);
                        }
                        yield new Type.TInt();
                    }
                    case "==", "!=", "<", "<=", ">", ">=" -> {
                        if (!(lt instanceof Type.TInt)) {
                            throw new TypeError("comparison " + b.op() + " requires int, got " + lt);
                        }
                        yield new Type.TInt();
                    }
                    default -> throw new TypeError("unknown op: " + b.op());
                };
            }
            case Expr.Call c -> {
                FunSig sig = funcs.get(c.name());
                if (sig == null) throw new TypeError("undefined function: " + c.name());
                if (c.args().size() != sig.params().size()) {
                    throw new TypeError(c.name() + ": expected " + sig.params().size()
                            + " args, got " + c.args().size());
                }
                for (int i = 0; i < c.args().size(); i++) {
                    Type at = checkExpr(c.args().get(i), env, funcs);
                    Type pt = sig.params().get(i);
                    if (!sameType(at, pt)) {
                        throw new TypeError(c.name() + ": arg type " + at + ", expected " + pt);
                    }
                }
                yield sig.retType();
            }
            case Expr.Index idx -> {
                Type at = checkExpr(idx.arr(), env, funcs);
                Type it = checkExpr(idx.idx(), env, funcs);
                if (!(it instanceof Type.TInt)) throw new TypeError("index must be int, got " + it);
                if (at instanceof Type.TList tl) yield tl.elem();
                throw new TypeError("cannot index non-list type: " + at);
            }
            case Expr.ListLit ll -> {
                if (ll.elems().isEmpty()) throw new TypeError("empty list literal is not supported");
                Type first = checkExpr(ll.elems().get(0), env, funcs);
                for (int i = 1; i < ll.elems().size(); i++) {
                    Type t = checkExpr(ll.elems().get(i), env, funcs);
                    if (!sameType(t, first)) {
                        throw new TypeError("list literal: mixed types " + first + " and " + t);
                    }
                }
                yield new Type.TList(first);
            }
        };
    }

    private static boolean sameType(Type a, Type b) {
        if (a instanceof Type.TInt && b instanceof Type.TInt) return true;
        if (a instanceof Type.TString && b instanceof Type.TString) return true;
        if (a instanceof Type.TList la && b instanceof Type.TList lb) {
            return sameType(la.elem(), lb.elem());
        }
        return false;
    }
}
